extensions-contrib/aliyun-oss-extensions/src/main/java/org/apache/druid/storage/aliyun/ObjectSummaryIterator.java

```diff
@@ -0,0 +1,163 @@
+  private final int maxListingLength;
+
+  private ListObjectsRequest request;
+  private ObjectListing result;
+  private Iterator<OSSObjectSummary> objectSummaryIterator;
+  private OSSObjectSummary currentObjectSummary;
+
+  ObjectSummaryIterator(
+      final OSS client,
+      final Iterable<URI> prefixes,
+      final int maxListingLength
+  )
+  {
+    this.client = client;
+    this.prefixesIterator = prefixes.iterator();
+    this.maxListingLength = maxListingLength;
+
+    prepareNextRequest();
+    fetchNextBatch();
+    advanceObjectSummary();
+  }
+
+  @Override
+  public boolean hasNext()
+  {
+    return currentObjectSummary != null;
+  }
+
+  @Override
+  public OSSObjectSummary next()
+  {
+    if (currentObjectSummary == null) {
+      throw new NoSuchElementException();
+    }
+
+    final OSSObjectSummary retVal = currentObjectSummary;
+    advanceObjectSummary();
+    return retVal;
+  }
+
+  private void prepareNextRequest()
+  {
+    final URI currentUri = prefixesIterator.next();
+    final String currentBucket = currentUri.getAuthority();
+    final String currentPrefix = OssUtils.extractKey(currentUri);
+
+    request = new ListObjectsRequest(currentBucket, currentPrefix, null, null, maxListingLength);
+  }
+
+  private void fetchNextBatch()
+  {
+    try {
+      result = OssUtils.retry(() -> client.listObjects(request));
+      request.setMarker(result.getNextMarker());
+      objectSummaryIterator = result.getObjectSummaries().iterator();
+    }
+    catch (OSSException e) {
+      throw new RE(
+          e,
+          "Failed to get object summaries from S3 bucket[%s], prefix[%s]; S3 error: %s",
+          request.getBucketName(),
+          request.getPrefix(),
+          e.getMessage()
+      );
+    }
+    catch (Exception e) {
+      throw new RE(
+          e,
+          "Failed to get object summaries from S3 bucket[%s], prefix[%s]",
+          request.getBucketName(),
+          request.getPrefix()
+      );
+    }
+  }
+
+  /**
+   * Advance objectSummaryIterator to the next non-placeholder, updating "currentObjectSummary".
+   */
+  private void advanceObjectSummary()
+  {
+    while (objectSummaryIterator.hasNext() || result.isTruncated() || prefixesIterator.hasNext()) {
+      while (objectSummaryIterator.hasNext()) {
+        currentObjectSummary = objectSummaryIterator.next();
+        // skips directories and empty objects
+        if (!isDirectoryPlaceholder(currentObjectSummary) && currentObjectSummary.getSize() > 0) {
+          return;
+        }
+      }
+
+      // Exhausted "objectSummaryIterator" without finding a non-placeholder.
+      if (result.isTruncated()) {
+        fetchNextBatch();
+      } else if (prefixesIterator.hasNext()) {
+        prepareNextRequest();
+        fetchNextBatch();
+      }
+    }
+
+    // Truly nothing left to read.
+    currentObjectSummary = null;
+  }
+
+  /**
+   * Checks if a given object is a directory placeholder and should be ignored.
```

Comment:
This should preserve the original javadoc about being adapted from jets3t code:

```
   * Adapted from org.jets3t.service.model.StorageObject.isDirectoryPlaceholder(). Does not include the check for
   * legacy JetS3t directory placeholder objects, since it is based on content-type, which isn't available in an
   * S3ObjectSummary.
```

Can you also add an entry noting the usage here? 
https://github.com/apache/druid/blob/master/LICENSE#L276

Reply:
There's no directory placeholder in aliyun OSS, so I made change to this function. As a result, there's no need to add a declaration to license

Reply:
@FrankChen021 The code is still derived from jets3t code, so it must be accounted for in LICENSE.

Reply:
@jon-wei  LICENSE has been updated in the lastest commit, please check it again.